modules/openy_gc_auth/modules/openy_gc_auth_yusa/src/Plugin/GCIdentityProvider/YUSA.php

```diff
@@ -0,0 +1,217 @@
+      '#type' => 'textfield',
+      '#default_value' => $config['auth_pass'],
+      '#required' => TRUE,
+    ];
+
+    $form['verification_type'] = [
+      '#title' => $this->t('Verification type'),
+      '#type' => 'radios',
+      '#options' => [
+        'membership_id' => $this->t('Membership ID'),
+        'email' => $this->t('Email'),
+      ],
+      '#default_value' => $config['verification_type'],
+      '#required' => TRUE,
+    ];
+
+    $form['id_field_text'] = [
+      '#title' => $this->t('ID field text'),
+      '#type' => 'textfield',
+      '#default_value' => $config['id_field_text'],
+      '#required' => TRUE,
+    ];
+
+    $form['enable_recaptcha'] = [
+      '#title' => $this->t('Enable ReCaptcha'),
+      '#description' => $this->t('Set to TRUE if you want ReCaptcha validation on login form.'),
+      '#type' => 'checkbox',
+      '#default_value' => $config['enable_recaptcha'],
+    ];
+
+    $form['verification'] = [
+      '#type' => 'details',
+      '#title' => $this->t('Email verification'),
+      '#open' => TRUE,
+    ];
+
+    $form['verification']['enable_email_verification'] = [
+      '#title' => $this->t('Enable Email verification'),
+      '#description' => $this->t('Set to TRUE if you want enable one-time login link sending to user email for verification.'),
+      '#type' => 'checkbox',
+      '#default_value' => $config['enable_email_verification'],
+    ];
+
+    $form['verification']['email_verification_link_life_time'] = [
+      '#type' => 'select',
+      '#title' => $this->t('Login link life time'),
+      '#description' => $this->t('Select the time period after which the verification link will expire.'),
+      '#default_value' => $config['email_verification_link_life_time'],
+      '#required' => TRUE,
+      '#options' => [
+        3600 => $this->t('1 hour'),
+        7200 => $this->t('2 hours'),
+        10800 => $this->t('3 hours'),
+        14400 => $this->t('4 hours'),
+        18000 => $this->t('5 hours'),
+        86400 => $this->t('1 day'),
+        172800 => $this->t('2 days'),
+      ],
+    ];
+
+    $form['verification']['email_verification_text'] = [
+      '#title' => $this->t('Email verification text'),
+      '#description' => $this->t('This text will be used in the email.'),
+      '#type' => 'text_format',
+      '#format' => 'full_html',
+      '#default_value' => $config['email_verification_text'],
+      '#required' => TRUE,
+    ];
+
+    $form['verification']['verification_message'] = [
+      '#title' => $this->t('Verification message'),
+      '#description' => $this->t('This text will be displayed in the application after unverified user tried to login.'),
+      '#type' => 'text_format',
+      '#format' => 'full_html',
+      '#default_value' => $config['verification_message'],
+      '#required' => TRUE,
+    ];
+
+    return $form;
+  }
+
+  /**
+   * Get list of Virtual Y roles.
+   */
+  private function getRoles() {
```

Comment:
I think we have to create a centralized service for it, since we have many places that are doing the same.
Here is another place with almost similar code:
https://github.com/fivejars/openy_gated_content/blob/master/src/Plugin/Field/FieldWidget/GatedContentPermissionsFieldWidget.php#L93